### DISCUSSION TITLE
How to use third party npm modules in vscode extension?

### DISCUSSION THREAD

**Discussion by vijitha311** [microsoft/vscode-discussions#10] [Extension Development QnA] [answered]
In my extension i am in need of using few third party esm type npm modules. For example, consider "fs-extra" npm module

I installed it through "npm i --save fs-extra" and imported (import * as fse from 'fs-extra') in extension.ts

I'm seeing below error in the import statement. So i added //@ts-ignore for that line alone and did build then extension activation failed with "Cannot use import statement outside a module"

![image](https://user-images.githubusercontent.com/53463910/171798806-0ab8500c-8a8f-49b6-a66e-693cd2fdf025.png)

My sample extension repository which i replicated the same issue for "fs-extra" - https://github.com/vijitha311/Sample_Vscode_Extension

**Comment by digitarald** [comment]
This is less an extension issue, but rather a TypeScript issue. See https://bobbyhadz.com/blog/typescript-could-not-find-a-declaration-file-for-module#fix---could-not-find-declaration-file-for-module-x-error

Have you tried running the command suggested in the tooltip, `npm i --save-dev @types/fs-extra` (or the fallback to create a `.d.ts` file?)

**Reply by vijitha311** [reply]
@digitarald I tried with "npm i --save-dev @types/fs-extra" option too. This makes the extension failed with "cannot find the module fs-extra" also now i checked the solutions mentioned in https://bobbyhadz.com/blog/typescript-could-not-find-a-declaration-file-for-module#fix---could-not-find-declaration-file-for-module-x-error, nothing worked for me. Error is still persisting

Yes this issue is more related to typescript.  Can you help me to resolve?

**Reply by ghost** [reply]
Do you know if there is an npm api?

**Comment by digitarald** [comment, marked as answer]
While the first is a TS problem, `cannot find the module fs-extra` would be an npm problem. It is also mentioned on the [same page](https://bobbyhadz.com/blog/javascript-cannot-find-module-express#solve---cannot-find-module-express-error-in-nodejs).

**Reply by vijitha311** [reply]
@digitarald i installed @types/fs-extra module in the way as mentioned in the page. If you don't mind, please check the shared github repository by me. It has both "fs-extra" and "@types/fs-extra" to check

**Reply by digitarald** [reply]
Testing your repo in codespaces, after running `npm i` everything just works:

<img width="846" alt="image" src="https://user-images.githubusercontent.com/8599/172172258-2495bcf7-57a4-4874-8e4e-b78f9868630e.png">

**Reply by vijitha311** [reply]
@digitarald Everything is perfect here. Extension will be failed whenever it gets activated. Please try executing hello world command, then extension will go inside activation method and fail in the import statement of fse

**Reply by burkeholland** [reply]
It looks like (according to your `package.json` that you don't have fse installed. If I run `npm i fs-extra`, your code works - although it fails because I don't have a D: drive.

**Comment by burkeholland** [comment]
Closing this as answered because it's past the window for response. Please re-open if necessary.

**Comment by umerkle** [comment]
I want to reopen this, because i added docx package
```
 "dependencies": {
    "docx": "^8.5.0"
  }
```
and tried to use it in a .ts file with
`import {Document} from "docx";`

But then vscode tells me:

> The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("docx")' call instead.
>   To convert this file to an ECMAScript module, change its file extension to '.mts', or add the field `"type": "module"` to 'c:/SCM/Git/GitLab/docxhelper/docxhelper/package.json'.ts(1479)

I started my extension as described in the "Getting Started". 
What can i do to be able to use docx in my extension?
